keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -248,11 +458,833 @@ Package `kuberuntime`
+alternative][phase-startup] is added in the future. When would the preStop hooks
+be executed now that there are several phases? At the beginning? If there are 5
+phases: 0, 1, 2, 3 and 4, how should the shutdown behaviour be? Run preStop
+hooks for containers marked in phase 0, then kill containers in phase 4, then
+preStop hooks for containers in phase 1, then kill containers in phase 3, etc.?
+Or only the preStop hooks for container 0 are called? Why?
+
+It seems confusing, there doesn't seem to be a clear answer on those cases nor
+what the use case would be for such semantics.
+
+[hook-delivery]: https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-delivery-guarantees
+[phase-startup]: #pre-defined-phases-for-container-startupshutdown-or-arbitrary-numbers-for-ordering
+
+##### Alternative 1: add a TerminationHook
+
+Add to containers a `TerminationHook` field. It will accept the same values as
+preStop hooks and will be called for any container that defines it when the pod
+is switching to a terminating state. These clear semantics seem easy to extend in
+the future.
+
+Then, instead of running preStop hooks twice for sidecar containers as it is now
+proposed, preStop hooks are run only once: just before stopping the
+corresponding containers.
+
+The shutdown sequence steps in this case would be:
+1. Run TerminationHook for _any_ container that defines it (sidecar or not)
+1. Run preStop hooks for non-sidecars
+1. Send SIGTERM for non-sidecars
+1. Run preStop hooks for sidecars
+1. Send SIGTERM for sidecars
+
+If containers want to take action when a pod is switching to terminating state,
+they should use the TerminationHook.
+
+The motivation for running the preStop hooks two times, then, can be implemented
+by adding the TerminationHook field to the service mesh container to drain
+connections.
+
+Furthermore, if a container type "OrderedPhases" with the semantics of the [explained
+alternative][phase-startup] is added in the future, the semantics are still
+clear: the TerminationHook will be called for _any_ container (irrespective of
+their phase) when the pod switches to a terminating state.
+
+##### Alternative 2: Do nothing
+
+Do not call preStop hooks for sidecars twice, just remove the first call this
+section discusses.
+
+Then, suggest users that need to know when the pod changes to a terminating state to
+integrate with the Kubernetes API and add a watcher for that.
+
+This might be inconvenient for several users and further conversation is needed
+to see if this is feasible. We are aware of some users that this will cause
+issues and will probably need to patch Kubernetes because of this. This could
+also result in a large increase in calls to the Kubernetes API server causing
+scalability issues.
+
+##### Suggestion
+
+Aim to use alternative 1, while in parallel we collect more feedback from the
+community.
+
+It will be nice to do nothing (alternative 2), but probably the first step
```

Comment:
> but probably the first step

It can always be added after the feedback. So we can go with do nothing and add exactly what is required later.

Reply:
Sure. If all agree on that, LGTM :)

Reply:
This was added by request of istio devs, though. Let's see what they say here: https://github.com/kubernetes/enhancements/pull/1913#discussion_r463293308

And then decide :)